crates/runtime/src/externref.rs

```diff
@@ -0,0 +1,445 @@
+///     let mut file = file.borrow_mut();
+///     writeln!(&mut file, "Hello, `VMExternRef`!")?;
+/// }
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct VMExternRef(NonNull<VMExternData>);
+
+#[repr(C)]
+struct VMExternData {
+    // Implicit, dynamically-sized member that always preceded an
+    // `VMExternData`.
+    //
+    // value: [u8],
+    //
+    /// The reference count for this `VMExternData` and value. When it reaches
+    /// zero, we can safely destroy the value and free this heap
+    /// allocation. This is an `UnsafeCell`, rather than plain `Cell`, because
+    /// it can be modified by compiled Wasm code.
+    ///
+    /// Note: this field's offset must be kept in sync with
+    /// `wasmtime_environ::VMOffsets::vm_extern_data_ref_count()` which is
+    /// currently always zero.
+    ref_count: UnsafeCell<usize>,
+
+    /// Always points to the implicit, dynamically-sized `value` member that
+    /// precedes this `VMExternData`.
+    value_ptr: NonNull<dyn Any>,
+}
+
+impl Clone for VMExternRef {
+    #[inline]
+    fn clone(&self) -> VMExternRef {
+        self.extern_data().increment_ref_count();
+        VMExternRef(self.0)
+    }
+}
+
+impl Drop for VMExternRef {
+    #[inline]
+    fn drop(&mut self) {
+        let data = self.extern_data();
+        data.decrement_ref_count();
+        if data.get_ref_count() == 0 {
+            // Drop our live reference to `data` before we drop it itself.
+            drop(data);
+            unsafe {
+                VMExternData::drop_and_dealloc(self.0);
+            }
+        }
+    }
+}
+
+impl VMExternData {
+    unsafe fn drop_and_dealloc(mut data: NonNull<VMExternData>) {
+        // Note: we introduce a block scope so that we drop the live
+        // reference to the data before we free the heap allocation it
+        // resides within after this block.
+        let (alloc_ptr, layout) = {
+            let data = data.as_mut();
+            debug_assert_eq!(data.get_ref_count(), 0);
+
+            // Same thing, but for the dropping the reference to `value` before
+            // we drop it itself.
+            let layout = {
+                let value = data.value_ptr.as_ref();
+
+                let value_size = mem::size_of_val(value);
+                let value_align = mem::align_of_val(value);
+
+                let extern_data_size = mem::size_of::<VMExternData>();
+                let extern_data_align = mem::align_of::<VMExternData>();
+
+                let value_and_padding_size = round_up_to_align(value_size, extern_data_align)
+                    .unwrap_or_else(|| unreachable!());
+
+                let alloc_align = std::cmp::max(value_align, extern_data_align);
+                let alloc_size = value_and_padding_size + extern_data_size;
+
+                debug_assert!(Layout::from_size_align(alloc_size, alloc_align).is_ok());
+                Layout::from_size_align_unchecked(alloc_size, alloc_align)
```

Comment:
This looks like it's sharing a lot of code with allocation below, could that be refactored out to a shared function?